client/etcd/client.go

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package etcd
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/m3db/m3cluster/client"
+	"github.com/m3db/m3cluster/kv"
+	etcdKV "github.com/m3db/m3cluster/kv/etcd"
+	"github.com/m3db/m3cluster/services"
+	sdClient "github.com/m3db/m3cluster/services/client"
+	"github.com/m3db/m3cluster/services/heartbeat"
+	etcdHeartbeat "github.com/m3db/m3cluster/services/heartbeat/etcd"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/log"
+	"github.com/uber-go/tally"
+)
+
+const (
+	keyFormat       = "%s/%s"
+	cacheFileFormat = "%s-%s.json"
+	kvPrefix        = "_kv"
+)
+
+type newETCDClientFn func(endpoints []string) (*clientv3.Client, error)
+
+// NewConfigServiceClient returns a ConfigServiceClient
+func NewConfigServiceClient(opts Options) client.Client {
+	return &csclient{
+		opts:    opts,
+		kvScope: opts.InstrumentOptions().MetricsScope().Tagged(map[string]string{"config_service": "kv"}),
```

Comment:
nit: could do `scope := opts.InstrumentOptions().MetricsScope()` and reuse `scope` to make this slightly shorter